aws-rds-cfn-common/src/main/java/software/amazon/rds/common/handler/Tagging.java

```diff
@@ -184,56 +185,52 @@ private static RemoveTagsFromResourceRequest removeTagsFromResourceRequest(
+        This is done to ensure backward compatibility. */
         if (tagsToAdd.getResourceTags().isEmpty() && tagsToRemove.getResourceTags().isEmpty()) {
-            return softFailErrorRuleSet;
+            return stackTagsErrorRuleSet;
         }
-        return hardFailErrorRuleSet;
+        return resourceTagsErrorRuleSet;
     }
 
-    public static <M, C extends TaggingContext.Provider> ProgressEvent<M, C> safeCreate(
+    public static <M, C extends TaggingContext.Provider> ProgressEvent<M, C> createWithUnauthorizedTagging(
             final AmazonWebServicesClientProxy proxy,
             final ProxyClient<RdsClient> rdsProxyClient,
             final HandlerMethod<M, C> handlerMethod,
             final ProgressEvent<M, C> progress,
             final Tagging.TagSet allTags
     ) {
-        return progress.then(p -> {
-            final C context = p.getCallbackContext();
-            if (context.getTaggingContext().isSoftFailTags()) {
-                return p;
+        final ProgressEvent<M, C> allTagsResult = handlerMethod.invoke(proxy, rdsProxyClient, progress, allTags);
+        if (allTagsResult.isFailed()) {
+            if (isUnauthorizedTaggingFailure(allTagsResult, allTags)) {
+                allTagsResult.setErrorCode(HandlerErrorCode.UnauthorizedTaggingOperation);
             }
-            final ProgressEvent<M, C> allTagsResult = handlerMethod.invoke(proxy, rdsProxyClient, p, allTags);
-            if (allTagsResult.isFailed()) {
-                if (HandlerErrorCode.AccessDenied.equals(allTagsResult.getErrorCode())) {
-                    context.getTaggingContext().setSoftFailTags(true);
-                    return ProgressEvent.progress(allTagsResult.getResourceModel(), context);
-                }
-                return allTagsResult;
-            }
-            allTagsResult.getCallbackContext().getTaggingContext().setAddTagsComplete(true);
             return allTagsResult;
-        }).then(p -> {
-            final C context = p.getCallbackContext();
-            if (!context.getTaggingContext().isSoftFailTags()) {
-                return p;
-            }
-            final Tagging.TagSet systemTags = Tagging.TagSet.builder().systemTags(allTags.getSystemTags()).build();
-            return handlerMethod.invoke(proxy, rdsProxyClient, p, systemTags);
-        });
+        }
+        allTagsResult.getCallbackContext().getTaggingContext().setAddTagsComplete(true);
+        return allTagsResult;
+    }
+
+    private static <M, C extends TaggingContext.Provider> boolean isUnauthorizedTaggingFailure(final ProgressEvent<M, C> allTagsResult,
+                                                                                               final TagSet allTags) {
+        return HandlerErrorCode.AccessDenied.equals(allTagsResult.getErrorCode()) &&
+                        allTags.getResourceTags().isEmpty() &&
```

Comment:
Does the model differ only by resource tags on second execution?

Reply:
There won't be another execution following this PR. However, if you're referring to whether a second execution occurred when the resource tags were empty and an access denied error occurred, then the answer is yes.